### PR TITLE
Check for symlink in builtin volume

### PR DESCRIPTION
**Pull request by umohnani8**
Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

#### What type of PR is this?

 /kind bug

#### What this PR does / why we need it:
Check if a builtin volume is a symlink. If it is,
follow the symlink and ensure that the destination
exists.
Add tests for symlink and no symlink case.

#### How to verify it

Follow reproducer in https://github.com/containers/buildah/issues/2204#issue-575395639

#### Which issue(s) this PR fixes:

Fixes https://github.com/containers/buildah/issues/2204 and https://github.com/containers/buildah/issues/2249

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: run_linux.go [hunk 1]
```diff
@@ -359,26 +359,36 @@ func runSetupBuiltinVolumes(mountLabel, mountPoint, containerDir string, builtin
 			}
 			initializeVolume = true
 		}
-		stat, err := os.Stat(srcPath)
-		if err != nil {
-			if !os.IsNotExist(err) {
-				return nil, err
+		// Check if srcPath is a symlink
+		stat, statErr := os.Lstat(srcPath)
```
Comment (rhatdan) [inline run_linux.go:363]: nit: no reason not to just leave this as err.

**Review by rhatdan — APPROVED** [review]

**Comment by TomSweeneyRedHat** [comment]
New tests looks a little unhappy.
```
[+0414s] not ok 74 bud-builtin-volume-symlink
[+0414s] # (from function `die' in file ./helpers.bash, line 196,
[+0414s] #  from function `run_buildah' in file ./helpers.bash, line 183,
[+0414s] #  in test file ./bud.bats, line 739)
[+0414s] #   `run_buildah run $cid echo hello' failed
```

**Comment by umohnani8** [comment]
Yeah, I was trying to figure out why the containerized integration test was unhappy but the same test in the other suites are fine...

File: run_linux.go (hunk 1 above)
Comment (umohnani8) [reply run_linux.go:363]: the smoke test is unhappy when I use err, it claims one of the stat `err` is unused, which is not the case. Going to leave it as `statErr`.

File: chroot/run.go [hunk 2]
```diff
@@ -1161,10 +1162,21 @@ func setupChrootBindMounts(spec *specs.Spec, bundlePath string) (undoBinds func(
 			}
 		}
 		target := filepath.Join(spec.Root.Path, m.Destination)
-		if _, err := os.Stat(target); err != nil {
+		// Check if target is a symlink
+		stat, statErr := os.Lstat(target)
+		// If target is a symlink, follow the link and ensure the destination exists
+		if stat != nil && (stat.Mode()&os.ModeSymlink != 0) {
```
Comment (rhatdan) [inline chroot/run.go:1168]: If you change the statErr above to err, and then change this line to 

if err == nil && (stat.Mode()&os.ModeSymlink != 0) {

Everything will be happy including me.  :^)

File: run_linux.go [hunk 3]
```diff
@@ -359,26 +359,36 @@ func runSetupBuiltinVolumes(mountLabel, mountPoint, containerDir string, builtin
...
+		// If srcPath is a symlink, follow the link and ensure the destination exists
+		if stat != nil && (stat.Mode()&os.ModeSymlink != 0) {
```
Comment (rhatdan) [inline run_linux.go:365]: If you change the statErr above to err, and then change this line to

if err == nil && (stat.Mode()&os.ModeSymlink != 0) {

Everything will be happy including me. :^)

**Review by rhatdan — APPROVED** [review]

File: run_linux.go (hunk 3 above)
Comment (umohnani8) [reply run_linux.go]: Fixed

**Comment by umohnani8** [comment]
This is ready, @nalind @TomSweeneyRedHat @rhatdan PTAL

File: run_linux.go [hunk 4]
```diff
@@ -359,7 +359,17 @@ func runSetupBuiltinVolumes(mountLabel, mountPoint, containerDir string, builtin
 			}
 			initializeVolume = true
 		}
-		stat, err := os.Stat(srcPath)
+		// Check if srcPath is a symlink
+		stat, err := os.Lstat(srcPath)
+		// If srcPath is a symlink, follow the link and ensure the destination exists
+		if err == nil && stat != nil && (stat.Mode()&os.ModeSymlink != 0) {
+			srcPath, err = copier.Eval(mountPoint, volume, copier.EvalOptions{})
```
Comment (nalind) [inline run_linux.go:366]: Should that second parameter be `srcPath`?

Comment (umohnani8) [reply run_linux.go:366]: srcPath is the joined path of `mountPoint` and `volume` and copier.Eval takes the root and then path separately, that is why the second param is volume.

**Comment by rhatdan** [comment]
LGTM
@vrothberg @saschagrunert @giuseppe PTAL

File: chroot/run.go [hunk 5]
```diff
@@ -1161,7 +1162,18 @@ func setupChrootBindMounts(spec *specs.Spec, bundlePath string) (undoBinds func(
...
 		}
 		target := filepath.Join(spec.Root.Path, m.Destination)
-		if _, err := os.Stat(target); err != nil {
+		// Check if target is a symlink
+		stat, err := os.Lstat(target)
+		// If target is a symlink, follow the link and ensure the destination exists
+		if err == nil && stat != nil && (stat.Mode()&os.ModeSymlink != 0) {
+			target, err = copier.Eval(spec.Root.Path, m.Destination, copier.EvalOptions{})
+			if err != nil {
+				return nil, errors.Wrapf(err, "error evaluating symlink %q", target)
```
Comment (saschagrunert) [inline chroot/run.go:1171]:
```suggestion
				return nil, errors.Wrapf(err, "evaluating symlink %q", target)
```

File: run_linux.go [hunk 6]
```diff
@@ -359,7 +359,17 @@ func runSetupBuiltinVolumes(mountLabel, mountPoint, containerDir string, builtin
...
+			if err != nil {
+				return nil, errors.Wrapf(err, "error evaluating symlink %q", srcPath)
```
Comment (saschagrunert) [inline run_linux.go:368]:
```suggestion
				return nil, errors.Wrapf(err, "evaluating symlink %q", srcPath)
```

**Review by saschagrunert — APPROVED** [review]
Just two nits, otherwise LGTM

**Comment by TomSweeneyRedHat** [comment]
Other than @saschagrunert's comments LGTM

File: chroot/run.go (hunk 5 above)
Comment (umohnani8) [reply chroot/run.go]: Fixed

File: run_linux.go (hunk 6 above)
Comment (umohnani8) [reply run_linux.go]: Fixed

**Review by saschagrunert — APPROVED** [review]

**Review by giuseppe — APPROVED** [review]
LGTM

**Pull request merged**